pkg/operators/common/systemdcache.go

```diff
@@ -0,0 +1,219 @@
+// Copyright 2023 The Inspektor Gadget authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package common
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/coreos/go-systemd/v22/dbus"
+	"github.com/sirupsen/logrus"
+
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/container-utils/cgroups"
+)
+
+type SystemdSubInterface interface {
+	AddCgroup(string, uint64)
+	RemoveCgroup(string, uint64)
+}
+
+type SystemdCache struct {
+	dbusConn *dbus.Conn
+
+	idToUnitname sync.Map
+
+	subscribers []SystemdSubInterface
+	subMutex    sync.Mutex
+
+	exit           chan struct{}
+	ticker         *time.Ticker
+	tickerDuration time.Duration
+
+	useCount      int
+	useCountMutex sync.Mutex
+}
+
+var (
+	systemdCache *SystemdCache
+	systemdOnce  sync.Once
+)
+
+func GetSystemdCache() (*SystemdCache, error) {
+	systemdOnce.Do(func() {
+		systemdCache, err = newSystemdCache(1 * time.Second)
+	})
+	return systemdCache, err
+}
+
+func (cache *SystemdCache) Subscribe(sub SystemdSubInterface, publishOldEntries bool) {
+	logrus.Debugf("Subscribing to systemd cache")
+	cache.subMutex.Lock()
+	cache.subscribers = append(cache.subscribers, sub)
+	cache.subMutex.Unlock()
+
+	// Publish old entries after subscribing
+	// This ensures that the subscriber always has all entries that it needs
+	// It may recieve some entries twice because of that
```

Comment:
1. User subscribes to the systemdcache (now cgroup-hook) and wants the old entries aswell
2. Before it gets the old entries it gets interrupted
3. A new systemd service starts
3.1 User will be notified of this new systemd service
3.2 The systemd service will be entered into the currently known entries list
4. Now the interrupted process from 2. gets resumed
5. The code iterates through all currently known entries and publishes to the new subscriber (including the already published new systemd service

The same problem will happen when we return a list of the current systemd entries instead of this "republish" old entries